Downsample/Downsample.cpp

```diff
@@ -0,0 +1,196 @@
+float isnanmask(float);
+float nantozero(float);
+Zivid::PointCloud downsample(Zivid::PointCloud&, int);
+
+int main()
+{
+	try
+	{
+		Zivid::Application zivid;
+
+		std::string Filename = "Zivid3D.zdf";
+		std::cout << "Reading " << Filename << " point cloud" << std::endl;
+		Zivid::Frame frame = Zivid::Frame(Filename);
+
+		auto pointCloud = frame.getPointCloud();
+		auto downSamplingFactor = 4;
+
+		Zivid::PointCloud pointCloudDownsampled(pointCloud.height()/downSamplingFactor, pointCloud.width()/downSamplingFactor);
+
+		pointCloudDownsampled = downsample(pointCloud, downSamplingFactor);
+
+		std::cout << "Setting up visualization" << std::endl;
+		{
+			Zivid::CloudVisualizer vis;
+			zivid.setDefaultComputeDevice(vis.computeDevice());
+
+			std::cout << "Displaying the original point cloud" << std::endl;
+			vis.showMaximized();
+			vis.show(pointCloud);
+			vis.resetToFit();
+
+			std::cout << "Running the visualizer. Blocking until the window closes" << std::endl;
+			vis.run();
+		}
+		{
+			Zivid::CloudVisualizer vis;
+			zivid.setDefaultComputeDevice(vis.computeDevice());
+
+			std::cout << "Displaying the downsampled point cloud" << std::endl;
+			vis.showMaximized();
+			vis.show(pointCloudDownsampled);
+			vis.resetToFit();
+
+			std::cout << "Running the visualizer. Blocking until the window closes" << std::endl;
+			vis.run();
+		}
+	}
+
+	catch (const std::exception &e)
+	{
+		std::cerr << "Error: " << Zivid::toString(e) << std::endl;
+		return EXIT_FAILURE;
+	}
+}
+
+Eigen::MatrixXf sumline(Eigen::MatrixXf& matrix, int downSamplingFactor)
+{
+	Eigen::MatrixXf reshapedMatrix = matrix.reshaped(downSamplingFactor, (int)matrix.rows() * (int)matrix.cols() / downSamplingFactor);
+	
+	std::function<float(float)> nan_to_zero_wrap = nantozero;
+
+	reshapedMatrix = reshapedMatrix.unaryExpr(nan_to_zero_wrap);
+
+	return reshapedMatrix.colwise().sum().reshaped((int)matrix.rows() / downSamplingFactor, (int)matrix.cols()).transpose();
+}
+
+Eigen::MatrixXf gridsum(Eigen::MatrixXf& r, int downSamplingFactor)
+{
+	return sumline(sumline(r, downSamplingFactor), downSamplingFactor);
+}
+
+float myround(float x)
+{
+	return std::round(x);
+}
+
+float isnanmask(float x)
+{
+	if (std::isnan(x))
+	{
+		return 0;
+	}
+	else
+	{
+		return 1;
+	}
+}
+
+float nantozero(float x)
+{
+	if (std::isnan(x))
+	{
+		return 0;
+	}
+	else
+	{
+		return x;
+	}
+}
+
+Zivid::PointCloud downsample(Zivid::PointCloud& pointCloud, int downSamplingFactor)
+{
+	/*
+	Function for downsampling a Zivid point cloud. The downsampling factor represents the denominator of a fraction that represents the size of the downsampled point cloud relative to the original point cloud, e.g. 2 - one-half,  3 - one-third, 4 one-quarter, etc.
+	*/
+
+	if ((downSamplingFactor % 2 != 0) || (pointCloud.height() % downSamplingFactor) || (pointCloud.width() % downSamplingFactor))
+	{
+		throw std::invalid_argument("Downsampling factor has to have one of the following values: 2, 3, 4, 5, 6.");
```

Comment:
This does not seem to be what the `if` checks for? It looks like it checks that the factor is a multiple of two, and that the point cloud height and width are multiples of the factor.

The exception message has to reflect the actual requirement.

Does it really have to be a multiple of two? If so, I suggest splitting the `if` in two checks. One for checking whether it's a multiple of two, with an exception message saying just that, and one for checking that the dimensions are multiples of the factor, with an exception message which includes the dimensions and the factor.